tests/integration/test_mhtml_e2e.py

```diff
@@ -0,0 +1,107 @@
+    basedir = os.path.dirname(__file__)
+    datadir = os.path.join(basedir, 'data', 'downloads', 'mhtml')
+    files = os.listdir(datadir)
+    return files
+
+
+def normalize_line(line):
+    line = line.rstrip('\n')
+    line = re.sub('boundary="---=_qute-[0-9a-f-]+"',
+                  'boundary="---=_qute-UUID"', line)
+    line = re.sub('^-----=_qute-[0-9a-f-]+$', '-----=_qute-UUID', line)
+    line = re.sub(r'localhost:\d{1,5}', 'localhost:(port)', line)
+    return line
+
+
+class DownloadDir:
+
+    """Abstraction over a download directory."""
+
+    def __init__(self, tmpdir):
+        self._tmpdir = tmpdir
+        self.location = str(tmpdir)
+
+    def read_file(self):
+        files = self._tmpdir.listdir()
+        assert len(files) == 1
+
+        with open(str(files[0]), 'r', encoding='utf-8') as f:
+            return f.readlines()
+
+    def compare_mhtml(self, filename):
+        with open(filename, 'r', encoding='utf-8') as f:
+            expected_data = [normalize_line(line) for line in f]
+        actual_data = self.read_file()
+        actual_data = [normalize_line(line) for line in actual_data]
+        assert actual_data == expected_data
+
+
+@pytest.fixture
+def download_dir(tmpdir):
+    return DownloadDir(tmpdir)
+
+
+@pytest.mark.parametrize('test_name', collect_tests())
+def test_mhtml(test_name, download_dir, quteproc, httpbin):
+    quteproc.set_setting('storage', 'download-directory',
+                         download_dir.location)
+    quteproc.set_setting('storage', 'prompt-download-directory', 'false')
+
+    test_dir = os.path.join(os.path.abspath(os.path.dirname(__file__)),
+                            'data', 'downloads', 'mhtml', test_name)
+    test_path = 'data/downloads/mhtml/{}'.format(test_name)
+
+    quteproc.open_path('{}/{}.html'.format(test_path, test_name))
+    download_dest = os.path.join(download_dir.location,
+                                 '{}-downloaded.mht'.format(test_name))
+    httpbin.after_test()
+    quteproc.send_cmd(':download --mhtml --dest "{}"'.format(download_dest))
+    quteproc.wait_for(category='downloads', module='mhtml',
+                      function='finish_file',
+                      message='File successfully written.')
+
+    expected_file = os.path.join(test_dir, '{}.mht'.format(test_name))
+    download_dir.compare_mhtml(expected_file)
+
+    with open(os.path.join(test_dir, 'requests'), encoding='utf-8') as f:
+        expected_requests = []
+        for line in f:
+            if line.startswith('#'):
+                continue
+            path = '/{}/{}'.format(test_path, line.strip())
+            expected_requests.append(httpbin.ExpectedRequest('GET', path))
+
+    actual_requests = httpbin.get_requests()
+    # Requests are unorderable, we need to convert to ExpectedRequests
+    actual_requests = map(httpbin.ExpectedRequest.from_request,
+                          actual_requests)
+    assert sorted(actual_requests) == sorted(expected_requests)
```

Comment:
Using `collections.Counter` and making `ExpectedRequest` hashable (which imo is a logical thing to do) we can get nice diffs:

> E         Omitting 8 identical items, use -v to show
> E         Differing items:
> E         {ExpectedRequest(verb='GET', path='/data/downloads/mhtml/complex/background.png'): 1} != {ExpectedRequest(verb='GET', path='/data/downloads/mhtml/complex/background.png'): 2}
> E         Use -v to get the full diff

Which I find easier to understand than the list difference.